src/codegen/interpreter/bytecode_builder.cpp

```diff
@@ -0,0 +1,1830 @@
+  else if (type == code_context_.float_type_)
+    return BytecodeFunction::GetOpcodeFromId(id + 4);
+  else if (type == code_context_.double_type_)
+    return BytecodeFunction::GetOpcodeFromId(id + 5);
+  else
+    throw NotSupportedException("llvm type not supported: " +
+                                CodeGen::Dump(type));
+}
+
+Opcode BytecodeBuilder::GetOpcodeForTypeIntTypes(Opcode untyped_op,
+                                                 llvm::Type *type) const {
+  index_t id = BytecodeFunction::GetOpcodeId(untyped_op);
+
+  // This function highly depends on the macros in bytecode_instructions.def!
+
+  if (type == code_context_.bool_type_ || type == code_context_.int8_type_)
+    return BytecodeFunction::GetOpcodeFromId(id + 0);
+  else if (type == code_context_.int16_type_)
+    return BytecodeFunction::GetOpcodeFromId(id + 1);
+  else if (type == code_context_.int32_type_)
+    return BytecodeFunction::GetOpcodeFromId(id + 2);
+  else if (type == code_context_.int64_type_ ||
+           type == code_context_.char_ptr_type_ || type->isPointerTy())
+    return BytecodeFunction::GetOpcodeFromId(id + 3);
+  else
+    throw NotSupportedException("llvm type not supported: " +
+                                CodeGen::Dump(type));
+}
+
+Opcode BytecodeBuilder::GetOpcodeForTypeFloatTypes(Opcode untyped_op,
+                                                   llvm::Type *type) const {
+  index_t id = BytecodeFunction::GetOpcodeId(untyped_op);
+
+  // This function highly depends on the macros in bytecode_instructions.def!
+
+  // float is missing!
+  if (type == code_context_.float_type_)
+    return BytecodeFunction::GetOpcodeFromId(id + 0);
+  if (type == code_context_.double_type_)
+    return BytecodeFunction::GetOpcodeFromId(id + 1);
+  else
+    throw NotSupportedException("llvm type not supported: " +
+                                CodeGen::Dump(type));
+}
+
+Opcode BytecodeBuilder::GetOpcodeForTypeSizeIntTypes(Opcode untyped_op,
+                                                     llvm::Type *type) const {
+  index_t id = BytecodeFunction::GetOpcodeId(untyped_op);
+
+  // This function highly depends on the macros in bytecode_instructions.def!
+
+  switch (code_context_.GetTypeSize(type)) {
+    case 1:
+      return BytecodeFunction::GetOpcodeFromId(id + 0);
+    case 2:
+      return BytecodeFunction::GetOpcodeFromId(id + 1);
+    case 4:
+      return BytecodeFunction::GetOpcodeFromId(id + 2);
+    case 8:
+      return BytecodeFunction::GetOpcodeFromId(id + 3);
+    default:
+      throw NotSupportedException("llvm type size not supported: " +
+                                  CodeGen::Dump(type));
+  }
+}
+
+Instruction &BytecodeBuilder::InsertBytecodeInstruction(
+    const llvm::Instruction *llvm_instruction, Opcode opcode,
+    const std::vector<index_t> &args) {
+  PELOTON_ASSERT(opcode != Opcode::undefined);
+
+  // calculate number of required instruction slots
+  const size_t number_instruction_slots =
```

Comment:
Is this formula described anywhere?

Reply:
It calculates the number of needed instruction slots á 8 Byte from the number of arguments. Every argument is 2 Byte and the Opcode is additional 2 Bytes (that's where the +1 comes from)

![grafik](https://user-images.githubusercontent.com/5031362/39479256-b21df9ea-4d32-11e8-8fd7-21954b3dbdad.png)

It uses this trick:
![grafik](https://user-images.githubusercontent.com/5031362/39479265-be6b3708-4d32-11e8-8628-c3860cd5dd19.png)


I'll add this explanation in the code and change `2` to `sizeof(uint16_t)` to make it uniform. 
(Great, now I have some fancy formulas to put in my thesis)

Reply:
Thanks for the explanation. 

Yes, I'm aware of the trick, we use it in our bitmaps. Thinking about it now, it might make sense to put this into a generic `MathUtil::ceil()` function?

Reply:
I created a file `src/include/util/math_util.h` with the function `size_t MathUtil::DivRoundUp(size_t numerator, size_t denominator)` and added it in the bytecode builder and the null bitmap.